allennlp/nn/beam_search.py

```diff
@@ -293,18 +304,34 @@ def new_step(
             for key, state_tensor in state.items():
                 if state_tensor is None:
                     continue
-                _, *last_dims = state_tensor.size()
-                # shape: (batch_size, beam_size, *)
-                expanded_backpointer = backpointer.view(
-                    batch_size, self.beam_size, *([1] * len(last_dims))
-                ).expand(batch_size, self.beam_size, *last_dims)
-
-                # shape: (batch_size * beam_size, *)
-                state[key] = (
-                    state_tensor.reshape(batch_size, self.beam_size, *last_dims)
-                    .gather(1, expanded_backpointer)
-                    .reshape(batch_size * self.beam_size, *last_dims)
-                )
+                if state_tensor.dim() == 3 and key in ["decoder_hidden", "decoder_context"]:
```

Comment:
```suggestion
                multilayer_decoder = state_tensor.dim() == 3 and key in {"decoder_hidden", "decoder_context"}
                if multilayer_decoder:
```
It's functionally the same, but it makes it a little clearer what this `if` statement does.

Reply:
The same would work for the other place where you do this.

Reply:
@dirkgr  Thanks for your advice. 
Then main question is : only the beam_search need the state shape like `(batch,...)`.  
We could reshape state here or **before(and after) the decoder layer in all models** . I prefer the first one, slightly change the code rather than adjust all models' decoder.